### DISCUSSION TITLE
Video Speed

### DISCUSSION THREAD

**Discussion by SintelixSimonP** [emgucv/emgucv#499] [General]
Hi
I am attempting to write a media player for viewing .mp4 videos.
I am attempting to build in VB.NET using emgu.CV v3.101 (NUGet)

App is working, but the speed of the video is very slow.

I have attempted with both a Async Function call with an Await Task.Delay based on the FPS
and also a Timer with the tick time based on FPS.

The video always runs very slow.

Does any have any functioning .NET example of how to play a video at full speed ?

Thanks

**Comment by emgucv** [comment]
Have you tried to use the latest v4.5.1 release?

What is the resolution of the video and what codec is it using? What is your computer's configuration (CPU, memory, graphic card)?

Without the Task.Delay, what is the FPS for reading? What is the expected FPS?

Is there any additional processing used? Or is it just reading and rendering?

**Reply by SintelixSimonP** [reply]
Hi
My original tests were with v3, as I was following a tutorial that I converted from C# to VB.net.
I have updated today (with some challenges!) to the latest 4.5.1 release and it is still very slow.

my Laptop is a decent machines
DELL G3, i7 Processor, 16GB RAM, NVIDIA GeForce GTX 1640 Graphics Card.

The video resolution is high 4040w x 1110h, but normal 4k Video's have the same problem.
Its a very Simple app, I must be doing something wrong :(

Sample Code Snippets- 
-----------------------------------------------------------------------------------------------------------------------------------
Menu Item  to load the file
 Private Sub OpenToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles OpenToolStripMenuItem.Click
        Dim ofd = New OpenFileDialog()
        ofd.Filter = "Video Files (*.mp4, *.flv)| *.mp4;*.flv"

        If ofd.ShowDialog = DialogResult.OK Then

            videocapture = New VideoCapture(ofd.FileName, 0)

            TotalFrames = Convert.ToInt32(VideoCapture.GetCaptureProperty(Emgu.CV.CvEnum.CapProp.FrameCount))
            FPS = Convert.ToInt32(VideoCapture.GetCaptureProperty(Emgu.CV.CvEnum.CapProp.Fps))
            IsPlaying = True
            CurrentFrame = New Mat()
            CurrentFrameNo = 0
            'TrackBar1.Minimum = 0
            TrackBar1.Maximum = TotalFrames - 1
            'TrackBar1.Value = 1
            PlayVideoAsync()

        End If
    End Sub
----------------------------------------------------------------------------------------
Function to Play Video (I have added this into a timer also)
 Private Async Function PlayVideoAsync() As Task
        If (videocapture Is Nothing) Then
            ' Rem - do nothing
        Else
            Try
                While IsPlaying = True And CurrentFrameNo < TotalFrames
                    videocapture.SetCaptureProperty(Emgu.CV.CvEnum.CapProp.PosFrames, CurrentFrameNo)
                    'videocapture.Read(CurrentFrame)
                    videocapture.Retrieve(CurrentFrame, 0)

                    'PictureBox1.Image = videocapture.QueryFrame().bitmap
                    ImageBox1.Image() = videocapture.QueryFrame()
                    TrackBar1.Value = CurrentFrameNo
                    CurrentFrameNo += 1
                    Await Task.Delay(1000 / FPS)
                End While

            Catch ex As Exception
                MessageBox.Show(ex.Message)
            End Try


        End If
    End Function

**Reply by emgucv** [reply]
In your code:

> `Await Task.Delay(1000 / FPS)`

With that much delay, you are basically assuming the decoding take 0 milliseconds. 

I would expect close to 90-100% CPU time when decoding 4K video using software decoding in Open CV. 

Can you also answer these questions:

> What is the video codec? 
> Without the Task.Delay, what is the FPS for reading? What is the expected FPS?

**Comment by SintelixSimonP** [comment]
Hi 

FPS = 30FPS  there for delay is 1000/30
I have tried longer delays - but there is no difference

Full Video Info
Codec - H264 : MPEG-4
Video Resolution  - 4040x110
Frame Rate - 30FPS

Video can be played easy enough in normal Media Players (VLC etc)
Sample video can be seen here
https://files.sintelix.net/index.php/s/4pcMEB55OMVI60F


Not sure what you mean by:
Without the Task.Delay, what is the FPS for reading? 

as without the task delay, the function will not return.

Thanks for the help - any advice welcome.

**Comment by emgucv** [comment]
Replace 

`Await Task.Delay(1000 / FPS)`

with 

`Await Task.Delay(0)`

Will you be able to get more than 30FPS? If so, your cpu is capable to render it at 30FPS with Emgu CV.

The next step is to measure how much time it took to do everything else (besides of the await call) in the while loop. 

If you measure the time for retriving a frame is "frame_processing_time_milliseconds", update your code to:
`Await Task.Delay( (1000 / FPS) - frame_processing_time_milliseconds )`

This way every step in the while loop took exactly (1000/FPS) milliseconds.

**Reply by emgucv** [reply]
Btw, you mention the video resultion is 4040 x 1110 in the earlier post, and mention again that the resolution is 4040x110 in a later post. Which one is the correct resolution? It is 10x size differences between the two resolutions.
